### PR TITLE
[spring-boot-hw-book-store-exception-handler] added dto fields validation and custom exception handler

### PR DESCRIPTION
**Pull request by Andriy-Hrynevetskyi**

### CODE REVIEW

**Review by DevIgork — APPROVED** [review]
LGTM

File: src/main/java/bookstore/dto/CreateBookRequestDto.java [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
 
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
 import java.math.BigDecimal;
 import lombok.Data;
 
 @Data
 public class CreateBookRequestDto {
+    private static final String TITLE_ERR_MSG = "Title field can't be null or empty";
```
Comment (vovasalyha) [inline src/main/java/bookstore/dto/CreateBookRequestDto.java]: IMO the messages can be inlined here for better cohesion, but it's up to you.

File: src/main/java/bookstore/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+    private static final String AUTHOR_ERR_MSG = "Author field can't be null or empty";
+    private static final String ISBN_ERR_MSG = "ISBN field can't be null or empty";
+    private static final String PRICE_ERR_MSG = "Price can't be null or less than 0";
+    @NotBlank(message = TITLE_ERR_MSG)
     private String title;
```
Comment (vovasalyha) [inline src/main/java/bookstore/dto/CreateBookRequestDto.java:15]: I would also validate a maximum size of text fields, to not exceed the DB column restrictions.

File: src/main/java/bookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -1,13 +1,25 @@
...
+    @NotBlank(message = AUTHOR_ERR_MSG)
     private String author;
+    @NotBlank(message = ISBN_ERR_MSG)
     private String isbn;
+    @NotNull(message = PRICE_ERR_MSG)
+    @Min(0)
```
Comment (vovasalyha) [inline src/main/java/bookstore/dto/CreateBookRequestDto.java]: You can use `Positive`

File: src/main/java/bookstore/dto/CreateBookRequestDto.java [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
+    private static final String TITLE_ERR_MSG = "Title field can't be null or empty";
+    private static final String AUTHOR_ERR_MSG = "Author field can't be null or empty";
+    private static final String ISBN_ERR_MSG = "ISBN field can't be null or empty";
+    private static final String PRICE_ERR_MSG = "Price can't be null or less than 0";
+    @NotBlank(message = TITLE_ERR_MSG)
     private String title;
+    @NotBlank(message = AUTHOR_ERR_MSG)
     private String author;
+    @NotBlank(message = ISBN_ERR_MSG)
     private String isbn;
```
Comment (vovasalyha) [inline src/main/java/bookstore/dto/CreateBookRequestDto.java:24]: ISBN has other validation rules (like fixed length and format)

File: src/main/java/bookstore/repository/book/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -17,7 +16,7 @@ public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
 
     @Override
     public Specification<Book> build(BookSearchParameters searchParameters) {
-/*        Specification<Book> spec = Specification.where(null);
+        Specification<Book> spec = Specification.where(null);
```
Comment (vovasalyha) [inline src/main/java/bookstore/repository/book/BookSpecificationBuilder.java]: Please try to not include not related changes to PRs.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Looks good, but see a couple of comments. Also, resolve all merge conflicts and don't include unrelated changes to PR

File: src/main/java/bookstore/dto/CreateBookRequestDto.java (hunk 4 above)
Comment (d1sam) [reply src/main/java/bookstore/dto/CreateBookRequestDto.java:24]: ISBN has fixed format, you can use [Hibernate validators](https://www.baeldung.com/hibernate-validator-constraints) or rewrite your regexp by [this](https://howtodoinjava.com/java/regex/java-regex-validate-international-standard-book-number-isbns/) guide

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
```
Comment (d1sam) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: make it as a constant

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (d1sam) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: make it as a constant

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            String message = e.getDefaultMessage();
+            return field + " " + message;
+        }
+        return e.getDefaultMessage();
+    }
+
+    @ExceptionHandler({EntityNotFoundException.class})
+    protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
```
Comment (d1sam) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: make it as a constant

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        body.put("status", HttpStatus.NOT_FOUND);
```
Comment (d1sam) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: make it as a constant

File: src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        body.put("message", ex.getMessage());
```
Comment (d1sam) [inline src/main/java/bookstore/exception/CustomGlobalExceptionHandler.java]: make it as a constant

**Review by d1sam — CHANGES_REQUESTED** [review]
Some mistakes still exists and can be rewritten better

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**